### PR TITLE
fix(inward): include room-linked timed item charges in bill totals

### LINKED ISSUE #23118: Room-linked dynamic Timed Item charges never reach the interim/final bill
## Summary

Time-based services ("Timed Items") that are dynamically linked to a room's
fee configuration (`RoomFacilityTimedItem`) are correctly snapshotted and
priced against the patient's room stay, but the priced amount never reaches
the interim bill or the final bill. The charge is calculated, persisted, and
visible only in the "Patient Room Details" room-stay breakdown — it is
silently dropped everywhere billing actually happens, so the patient is
under-billed for that service with no error or warning anywhere in the UI.

## Environment

- Branch/commit: `development` @ `48861f82c17164126c09ad4dba1100aa1a3eb21a`
- Department: Inward, Galle Co Operative Hospital (local `coop` test DB)
- User role: buddhika (admin/dev, local testing)
- Test admission: `BHT/56755` ("Mr. Demo Test Patient", pre-existing local
  demo/test data, not a real patient), room `Room 100 wm`

## Steps to reproduce

1. Create (or use an existing) `TimedItem` whose "Inward Charge Category" is
   **not** one of the 7 charge types that already have dedicated fields on
   `RoomFacilityCharge` (Room/Maintenance/MO/Nursing/Linen/Administration/
   Medical Care) — e.g. category "Nebulisation". `inward_timed_item.xhtml`.
2. Give it a fee tier via **Manage Service Fees**
   (`inward_timed_item_fee.xhtml`) — e.g. Rs 500 / 1 hour block, Hospital Fee.
3. On **Room Fees → Duration-based Timed Item Charges**
   (`inward_room_facility.xhtml`), select a room (e.g. "Room 100 wm") and add
   the timed item from step 1 as a linked, dynamically-added charge for that
   room. This creates a `RoomFacilityTimedItem` link.
4. Open (or create) a patient admission whose current room uses that room
   facility charge, then open **Patient Room Details**
   (`inward_patient_room_details.xhtml`) and expand the room row → the
   "Duration-based Timed Item Charges" sub-table correctly lists the item and
   (once recalculated) a non-zero **Calculated/Net** amount.
5. Open the **Interim Bill** (`inward_bill_intrim.xhtml`) for the same
   admission.

## Expected

The priced timed-item charge from step 4 is included in the interim bill's
"Charges" breakdown (under its own Inward Charge Category, e.g.
"Nebulisation") and added into the Gross Total / Total Charges. The same
should hold for the Final Bill generated from this screen after discharge.

## Actual

The charge does not appear anywhere on the interim bill:
- It has no row in the "Charges" breakdown table.
- The "Timed Service Details" tab (a separate, unrelated data source for
  ad-hoc billed timed services) shows "No records found."
- Gross Total / Total Charges do not include it at all.

In the reproduction, `PatientRoomTimedItemCharge` was correctly computed by
the app to **Rs 101,508.33** for the 203-hour test stay (verified directly in
the `PATIENTROOMTIMEDITEMCHARGE` table), while the interim bill's Gross Total
stayed at **Rs 38,499.80** — completely excluding it. This is not a rounding
or display issue; the money is never billed.

The same aggregation path (`BhtSummeryController.chargeItemTotals`) backs the
Final Bill page (`inward_bill_final.xhtml`) as well, so the final bill would
under-bill identically.

## Root cause

Two compounding gaps in `src/main/java/com/divudi/bean/inward/BhtSummeryController.java`
and `src/main/java/com/divudi/bean/inward/InwardBeanController.java`:

**1. Bill-totals aggregation only handles 7 hardcoded charge types**

- `InwardBeanController.getTimedItemChargeSumsBulk()` correctly sums
  `PatientRoomTimedItemCharge.calculatedCharge - discountCharge`, grouped by
  each linked `TimedItem`'s own `inwardChargeType`
  (`InwardBeanController.java:1563` area, ~line 1543-1568) — this part works
  and is generic over all charge types.
- But `BhtSummeryController.setKnownChargeTot()`
  (`BhtSummeryController.java:5114` onward) only copies that map into the
  bill's `chargeItemTotals` for a `switch` statement covering exactly
  `RoomCharges`, `MOCharges`, `NursingCharges`, `MaintainCharges`,
  `MedicalCareICU`, `AdministrationCharge`, `LinenCharges` (plus a handful of
  unrelated cases like `Medicine`, `ProfessionalCharge`, etc. — see
  `BhtSummeryController.java:5140-5197`). There is no `default` branch, so
  any `TimedItem` whose `inwardChargeType` is one of the ~150 *other*
  selectable categories (`inward_timed_item.xhtml`'s "Inward Charge Category"
  dropdown, driven by `EnumController.getInwardChargeTypesForSetting()`,
  which allows the full `InwardChargeType` enum) has its already-correctly-
  computed sum silently discarded — even though a `ChargeItemTotal` row for
  that exact category already exists in the list (created earlier in
  `createChargeItemTotals()` for every type in
  `getInwardChargeTypesForSetting()`), its `.total` is simply never set from
  `roomSums`.
- Real-world evidence this isn't a hypothetical edge case: the only
  currently-defined `TimedItem` in the local `coop` DB before this
  investigation ("Nebulization Test 22316") already uses charge category
  "Nebulisation" — not one of the 7 handled types.

**2. `snapshotTimedItems()` only runs at initial room admission**

- `InwardBeanController.snapshotTimedItems(PatientRoom, RoomFacilityCharge)`
  (`InwardBeanController.java:2695` area) is what creates the
  `PatientRoomTimedItemCharge` link rows from `RoomFacilityTimedItem`. It is
  only called from the two `savePatientRoom(...)` overloads used during
  admission (`InwardBeanController.java:2596-2632` and `2652-2687`).
- It is **never** called from `updateChargesForRoom()`
  (`BhtSummeryController.java:2111`, the "Update Charges from Room Rates"
  button on Patient Room Details) or any other recalculation path.
- Practical effect: if a timed item is linked to a room facility charge
  *after* patients are already admitted to a room using that facility charge
  (the realistic order of operations — you configure room charges once, then
  extend them over time as new services are added), those already-admitted
  patients never get the new `PatientRoomTimedItemCharge` row created at all,
  with no way to backfill it short of a fresh room (re)admission. Confirmed
  by reproduction: linking the timed item to `Room 100 wm`'s facility charge
  did not retroactively create a charge row for the already-admitted test
  patient; a manual DB insert (mirroring exactly what `snapshotTimedItems()`
  would have inserted) was needed to demonstrate gap #1 above.

## Evidence

Interim bill missing the Nebulisation charge

Screenshot shows the Interim Bill for `BHT/56755`: Gross Total Rs 38,499.80,
no "Nebulisation" line in the Charges table, and the "Timed Service Details"
tab reporting "No records found" — while the room's own
`PATIENTROOMTIMEDITEMCHARGE` row (verified via direct DB query) held a
correctly-calculated Rs 101,508.33 for the linked timed item.

### PR DESCRIPTION
**Pull request by buddhika75**
## Decisions made without approval

This PR was developed unattended (`dev-issue-unattended`) while the requesting
developer was away. Every judgment call below was made autonomously and
should be double-checked first:

1. **Filed the issue myself** from a live demonstration/investigation session
   (browser repro + code reading), before handing off to the unattended
   pipeline — see issue #23118 for the full root-cause write-up and evidence.
2. **Fix shape for `setKnownChargeTot()`**: chose to generalize the existing
   `roomSums` merge into a single loop (matching the additive pattern already
   used by `setServiceTotCategoryWise()`/`setTimedServiceTotCategoryWise()`
   two methods later in the same file) rather than adding an 8th, 9th, ...
   explicit `case` per newly-discovered charge category. This was picked
   because it's the smallest change that fixes the bug for *every* charge
   category (not just the one reproduced), and it's consistent with a pattern
   the codebase already uses twice — not a new design.
3. **Fix shape for the missing backfill**: chose to call the existing,
   already-idempotent `snapshotTimedItems()` from `updateChargesForRoom()`
   (the "Update Charges from Room Rates" button) rather than, e.g., adding a
   scheduled job or a new button. This was picked because it's the one
   existing user-facing action that already refreshes a room's charges from
   its `RoomFacilityCharge`, so it's the natural place to also catch up newly
   linked timed items, and it required no new UI.
4. **Test data**: reused an existing local demo/test admission (`BHT/56755`,
   a synthetic "Demo Test Patient", not a real patient) already admitted to
   an existing test room ("Room 100 wm"), and reused the one pre-existing
   `TimedItem` in the local DB ("Nebulization Test 22316", category
   "Nebulisation") rather than creating new fixtures, per the "prefer an
   existing record" rule.

## Summary

Room-linked dynamic Timed Item charges (`RoomFacilityTimedItem` →
`PatientRoomTimedItemCharge`) were correctly calculated and shown on the
Patient Room Details page, but never reached the interim or final bill
totals — the patient was silently under-billed for that charge.

Fixes #23118.

## Root cause

Two compounding gaps in `BhtSummeryController.java` — see the issue body for
the full trace with line numbers. In short:

1. `setKnownChargeTot()` only copied the bulk `roomSums` map into
   `chargeItemTotals` via a `switch` hardcoded to 7 specific
   `InwardChargeType`s. Any `TimedItem` using one of the ~150 *other*
   selectable charge categories had its already-computed charge silently
   dropped.
2. `snapshotTimedItems()` — which creates the `PatientRoomTimedItemCharge`
   link row a `RoomFacilityTimedItem` needs — was only called from the two
   admission-time `savePatientRoom()` overloads, never from any
   recalculation path, so linking a timed item to a room *after* a patient
   was already admitted to that room type never created a charge row for
   that patient.

## Fix

- `setKnownChargeTot()`: replaced the 7-case switch handling of `roomSums`
  with one generic loop — `roomSums.get(i.getInwardChargeType())` applied to
  every `ChargeItemTotal` — matching the additive pattern already used by
  `setServiceTotCategoryWise()`/`setTimedServiceTotCategoryWise()`. No
  behavior change for the 7 original types; every other type now flows
  through too.
- `updateChargesForRoom()` (the "Update Charges from Room Rates" action) now
  also calls the existing, already-idempotent `snapshotTimedItems(pr, rfc)`,
  backfilling newly-linked timed items for already-admitted patients.

## Verification

Reproduced against a local test admission (`BHT/56755`, synthetic demo
patient, Room 100 wm) with the existing "Nebulization Test 22316" timed item
(category "Nebulisation") linked to that room's fee configuration:

- **Before fix**: Gross Total `38,499.80`, no "Nebulisation" line anywhere on
  the interim bill, despite `PATIENTROOMTIMEDITEMCHARGE` correctly holding a
  computed `101,691.67` for it.
- **After fix** (via the app's own "Update Charges from Room Rates" → Interim
  Bill navigation, not a raw DB check): Gross Total `140,191.46`, with a
  dedicated `Nebulisation: 101,691.67` line — all other charge lines (Room
  Charges still exactly `30,600.00`, etc.) unchanged, confirming no
  double-counting.

Full before/after screenshots and detail are on the issue:
https://github.com/hmislk/hmis/issues/23118#issuecomment-5332872818

Build (`mvn clean package -DskipTests`) succeeded; local Payara redeploy
showed no exceptions in `server.log` for either action.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved accuracy when updating room and facility charges.
  * Charge totals are now calculated more consistently across supported charge types.
  * Timed charges are now reflected correctly after room charges are saved, including package-locked rooms.
  * Room-linked timed charges are included without replacing existing category totals.
  * Admission fees continue to be handled separately and accurately.

### CODE REVIEW

File: src/main/java/com/divudi/bean/inward/BhtSummeryController.java [hunk 1]
```diff
@@ -2129,6 +2129,7 @@ public void updateChargesForRoom(PatientRoom pr) {
         pr.setCurrentAdministrationCharge(rfc.getAdminstrationCharge());
         pr.setCurrentMedicalCareCharge(rfc.getMedicalCareCharge());
         getPatientRoomFacade().edit(pr);
+        getInwardBean().snapshotTimedItems(pr, rfc);
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/java/com/divudi/bean/inward/BhtSummeryController.java:2135]: Fixed in 4e0c27bc: `updateChargesForRoom()` now calls `snapshotTimedItems(pr, pr.getRoomFacilityCharge())` in the package-room branch too, before the early return, matching your suggested diff exactly. Re-verified end-to-end after the change with no regression.

File: src/main/java/com/divudi/bean/inward/BhtSummeryController.java [hunk 2]
```diff
@@ -5140,34 +5141,20 @@ private void setKnownChargeTot() {
...
+            }
+        }
+
         for (ChargeItemTotal i : chargeItemTotals) {
             switch (i.getInwardChargeType()) {
                 case AdmissionFee:
                     if (getPatientEncounter().getAdmissionType() != null) {
                         i.setTotal(getInwardBean().getAdmissionCharge(getPatientEncounter(), childPatientEncouters));
                     }
                     break;
```
[the remark this answers is not in the document]
Comment (buddhika75) [reply src/main/java/com/divudi/bean/inward/BhtSummeryController.java]: Fixed in 4e0c27bc: moved the generic `roomSums` merge to after the category-specific switch and made it additive (`i.setTotal(i.getTotal() + roomSum)`), matching your suggested diff. This stops the switch's plain-assignment cases (AdmissionFee/Medicine/CancelledReturnedMedicine/GeneralIssuing/ProfessionalCharge/DoctorAndNurses) from overwriting a room-linked timed item charge that happens to share one of those categories. Re-verified end-to-end with no regression on the original repro.
